### ISSUE TITLE
java.lang.AbstractMethodError: abstract method void com.google.android.exoplayer2.drm.DrmSessionManager.prepare()

### ISSUE THREAD

**Issue by sandarshnaroju** [google/ExoPlayer#8227]
### [REQUIRED] Searched documentation and issues
I have searched regarding this error in all issues of this repo, but nothing worked.
### [REQUIRED] Question
Currently, I am trying to write a bridge for exoplayer for react-native. I am seeing this error.   

### A full bug report captured from the device
```
2020-11-16 09:58:37.543 4494-4668/com.ulkatv E/AndroidRuntime: FATAL EXCEPTION: ExoPlayerImplInternal:Handler
    Process: com.ulkatv, PID: 4494
    java.lang.AbstractMethodError: abstract method "void com.google.android.exoplayer2.drm.DrmSessionManager.prepare()"
        at com.google.android.exoplayer2.source.dash.DashMediaSource.prepareSourceInternal(DashMediaSource.java:664)
        at com.google.android.exoplayer2.source.BaseMediaSource.prepareSource(BaseMediaSource.java:153)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.prepareInternal(ExoPlayerImplInternal.java:460)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:314)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
```
### Link to test content
My project configurations:

```
    compileSdkVersion 29

    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }

    defaultConfig {
        applicationId "com.ulkatv"
        minSdkVersion 23
        //noinspection ExpiredTargetSdkVersion
        targetSdkVersion 25
        versionCode 61
        versionName "0.6.1"
        multiDexEnabled true
    }
...

    implementation platform('com.google.firebase:firebase-bom:25.12.0')
    implementation 'com.google.firebase:firebase-analytics'
    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
    implementation 'com.google.android.gms:play-services-cast-framework:19.0.0'
    implementation 'com.google.android.exoplayer:exoplayer:2.11.8'

```

```
buildscript {
    ext {
        buildToolsVersion = "29.0.2"
        minSdkVersion = 19
        compileSdkVersion = 29
        targetSdkVersion = 29
    }
    repositories {
        google()
        jcenter()
    }
    dependencies {
        classpath("com.android.tools.build:gradle:3.5.4")
        classpath 'com.google.gms:google-services:4.3.4'
        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

```

**Comment by icbaker** [comment]
Errors like this are generally related to Java 8 support (the method it's complaining about is marked `default` on the `DrmSessionManager` interface, which is a language feature added in Java 8).

There's a [comment](https://github.com/google/ExoPlayer/issues/6801#issuecomment-574089568) here suggesting to try `android.enableDexingArtifactTransform=false` - did you try that?

If you'd like us to investigate further please provide a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) that demonstrates the problem in a way that we can build locally. 

This could be an Android Studio project on GitHub, or zipped up and sent to dev.exoplayer@gmail.com using a subject in the format "Issue #1234", where "#1234" should be replaced with your issue number. Please also update this issue to indicate you’ve done this.

**Comment by sandarshnaroju** [comment]
Just tried adding `android.enableDexingArtifactTransform=false ` in `gradle.properties`. It generated below error.

```
2020-11-16 15:52:52.351 10120-10308/com.ulkatv E/AndroidRuntime: FATAL EXCEPTION: ExoPlayerImplInternal:Handler
    Process: com.ulkatv, PID: 10120
    java.lang.AbstractMethodError: abstract method "java.lang.Class com.google.android.exoplayer2.drm.DrmSessionManager.getExoMediaCryptoType(com.google.android.exoplayer2.drm.DrmInitData)"
        at com.google.android.exoplayer2.source.dash.DashMediaPeriod.buildPrimaryAndEmbeddedTrackGroupInfos(DashMediaPeriod.java:667)
        at com.google.android.exoplayer2.source.dash.DashMediaPeriod.buildTrackGroups(DashMediaPeriod.java:505)
        at com.google.android.exoplayer2.source.dash.DashMediaPeriod.<init>(DashMediaPeriod.java:134)
        at com.google.android.exoplayer2.source.dash.DashMediaSource.createPeriod(DashMediaSource.java:686)
        at com.google.android.exoplayer2.MediaPeriodHolder.createMediaPeriod(MediaPeriodHolder.java:409)
        at com.google.android.exoplayer2.MediaPeriodHolder.<init>(MediaPeriodHolder.java:97)
        at com.google.android.exoplayer2.MediaPeriodQueue.enqueueNextMediaPeriodHolder(MediaPeriodQueue.java:155)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.maybeUpdateLoadingPeriod(ExoPlayerImplInternal.java:1547)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.updatePeriods(ExoPlayerImplInternal.java:1534)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:565)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:329)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
```

**Comment by sandarshnaroju** [comment]
I have created sample project that is producing the first error: [here](https://github.com/sandarshnaroju/videoplayer). Kindly look into it. Since my project is a react native one. Kindly locate the android folder and run it in android studio. I have added code exactly what I have tried without changing anything (Except credentials, tokens).

**Comment by icbaker** [comment]
Thanks for the project link. At 14.5k lines, I'm not sure it's a **minimal** example.

It doesn't build for me when I follow your instructions:
```
Settings file '[path to project]/videoplayer/android/settings.gradle' line: 3

A problem occurred evaluating settings 'videoplayer'.
> Could not read script '/[path to project]/videoplayer/node_modules/@react-native-community/cli-platform-android/native_modules.gradle' as it does not exist.
```

That `node_modules` directory doesn't seem to be present on GitHub.

**Comment by sandarshnaroju** [comment]
Hmm, Let me upload a pure android studio project instead of react-native one. Thanks for taking time.

**Comment by sandarshnaroju** [comment]
Sorry for late. Here is simple Android Project that is generating the error [Android Project](https://github.com/sandarshnaroju/VideoPlayer). Thanks for taking time.

**Comment by sandarshnaroju** [comment]
@icbaker any update on this?

**Comment by icbaker** [comment]
Thanks for the sample app, it builds but when deploying it fails with a `FileNotFoundException`:

```
2020-11-18 16:20:21.497 27575-27760/com.example.videoplayer E/ExoPlayerImplInternal: Source error
      com.google.android.exoplayer2.upstream.FileDataSource$FileDataSourceException: com.google.android.exoplayer2.upstream.FileDataSource$FileDataSourceException: java.io.FileNotFoundException: : open failed: ENOENT (No such file or directory)
        at com.google.android.exoplayer2.upstream.FileDataSource.open(FileDataSource.java:97)
        at com.google.android.exoplayer2.upstream.DefaultDataSource.open(DefaultDataSource.java:177)
        at com.google.android.exoplayer2.upstream.StatsDataSource.open(StatsDataSource.java:83)
        at com.google.android.exoplayer2.source.ProgressiveMediaPeriod$ExtractingLoadable.load(ProgressiveMediaPeriod.java:962)
        at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:415)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:920)
```

I suspect because the URI it's trying to open seems to be the empty string: 
https://github.com/sandarshnaroju/VideoPlayer/blob/f34745eee6db5573dcefafe7a797b55507f4a7aa/app/src/main/java/com/example/videoplayer/MainActivity.java#L133

**Comment by ShengjieGeng** [comment]
it seems like I have the similar error
below is the log
```

1 com.google.android.exoplayer2.ExoPlayerImpl$PlaybackInfoUpdate.lambda$run$0$ExoPlayerImpl$PlaybackInfoUpdate(ExoPlayerImpl.java:804)
--
2 com.google.android.exoplayer2.-$$Lambda$ExoPlayerImpl$PlaybackInfoUpdate$N_S5kRfhaRTAkH28P5luFgKnFjQ.invokeListener(lambda)
3 com.google.android.exoplayer2.BasePlayer$ListenerHolder.invoke(BasePlayer.java:182)
4 com.google.android.exoplayer2.ExoPlayerImpl.invokeAll(ExoPlayerImpl.java:845)
5 com.google.android.exoplayer2.ExoPlayerImpl.access$000(ExoPlayerImpl.java:43)
6 com.google.android.exoplayer2.ExoPlayerImpl$PlaybackInfoUpdate.run(ExoPlayerImpl.java:802)
7 com.google.android.exoplayer2.ExoPlayerImpl.notifyListeners(ExoPlayerImpl.java:736)
8 com.google.android.exoplayer2.ExoPlayerImpl.updatePlaybackInfo(ExoPlayerImpl.java:710)
9 com.google.android.exoplayer2.ExoPlayerImpl.handlePlaybackInfo(ExoPlayerImpl.java:652)
10 com.google.android.exoplayer2.ExoPlayerImpl.handleEvent(ExoPlayerImpl.java:595)
11 com.google.android.exoplayer2.ExoPlayerImpl$1.handleMessage(ExoPlayerImpl.java:127)
12 android.os.Handler.dispatchMessage(Handler.java:102)
13 android.os.Looper.loop(Looper.java:135)
14 android.app.ActivityThread.main(ActivityThread.java:5254)
15 java.lang.reflect.Method.invoke(Native Method)
16 java.lang.reflect.Method.invoke(Method.java:372)
17 com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
18 com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)


```

```
--------- beginning of main
--
111-17 18:45:03.946 1196 1196 E MtaSDK.CaughtExp: MTA has caught the following uncaught exception:
211-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp:
311-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: java.lang.AbstractMethodError: abstract method "void com.google.android.exoplayer2.Player$EventListener.onTimelineChanged(com.google.android.exoplayer2.Timeline, int)"
411-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.ExoPlayerImpl$PlaybackInfoUpdate.lambda$run$0$ExoPlayerImpl$PlaybackInfoUpdate(ExoPlayerImpl.java:804)
511-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.-$$Lambda$ExoPlayerImpl$PlaybackInfoUpdate$N_S5kRfhaRTAkH28P5luFgKnFjQ.invokeListener(lambda)
611-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.BasePlayer$ListenerHolder.invoke(BasePlayer.java:182)
711-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.ExoPlayerImpl.invokeAll(ExoPlayerImpl.java:845)
811-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.ExoPlayerImpl.access$000(ExoPlayerImpl.java:43)
911-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.ExoPlayerImpl$PlaybackInfoUpdate.run(ExoPlayerImpl.java:802)
1011-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.ExoPlayerImpl.notifyListeners(ExoPlayerImpl.java:736)
1111-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.ExoPlayerImpl.updatePlaybackInfo(ExoPlayerImpl.java:710)
1211-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.ExoPlayerImpl.handlePlaybackInfo(ExoPlayerImpl.java:652)
1311-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.ExoPlayerImpl.handleEvent(ExoPlayerImpl.java:595)
1411-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.google.android.exoplayer2.ExoPlayerImpl$1.handleMessage(ExoPlayerImpl.java:127)
1511-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at android.os.Handler.dispatchMessage(Handler.java:102)
1611-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at android.os.Looper.loop(Looper.java:135)
1711-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at android.app.ActivityThread.main(ActivityThread.java:5254)
1811-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at java.lang.reflect.Method.invoke(Native Method)
1911-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at java.lang.reflect.Method.invoke(Method.java:372)
2011-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
2111-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
2211-17 18:45:03.948 1196 1196 E MtaSDK.CaughtExp: App uncaught exception end.


```

**Comment by sandarshnaroju** [comment]
OK. I think I partially solved this issue. Earlier I loaded pallycon.aar as a module, looks that is the issue. I removed it as module, and added it in libs folder. And now the error disappeared. There is another problem, I am facing right now is, videos are playing when I set `drmsessionmanager` to the `ExoplayerFactory` (deprecated). But When I set the `drmsessionmanager` to media source it is showing blank page.
